File: src/main/java/ru/alepar/flr/http/HttpClient.java
package ru.alepar.flr.http;

import java.io.IOException;
import java.util.Map;

public interface HttpClient {
    String post(String url, Map<String, String> headers, Map<String, String> cookies, Map<String, String> data) throws IOException;

    String get(String url, Map<String, String> headers, Map<String, String> cookies, Map<String, String> data) throws IOException;

    byte[] execute(String url, Map<String, String> headers, Map<String, String> cookies, Map<String, String> data) throws IOException;
}

File: src/main/kotlin/io/alepar/flocal/download/HttpClient.kt
package io.alepar.flocal.download

import org.asynchttpclient.AsyncHttpClient
import org.asynchttpclient.Dsl
import java.util.concurrent.Semaphore

val httpClient: AsyncHttpClient = Dsl.asyncHttpClient(Dsl.config()
        .setMaxRequestRetry(3)
)
val httpSemaphore = Semaphore(concurrency)
fun download(url: String): String {
    httpSemaphore.acquire()
    try {
        return httpClient.prepareGet(url).execute().get().responseBody
    } finally {
        httpSemaphore.release()
    }
}